bugsnag-android-core/src/main/java/com/bugsnag/android/Journal.java

```diff
@@ -0,0 +1,170 @@
+ */
+public class Journal {
+    @NonNull
+    public final List<Command> commands;
+
+    public Journal() {
+        commands = new LinkedList<>();
+    }
+
+    /**
+     * Apply this journal to a document.
+     *
+     * This method goes through the journal entries one-by-one, applying their modification
+     * commands to the document. This is a destructive operation that will modify the document.
+     *
+     * Note: If a journal command replaces the root object, the resulting document will be a
+     * different object from the one passed in.
+     *
+     * @param document The document to apply to.
+     * @return The finished document (may be a different object from the one passed in).
+     */
+    public @NonNull Object applyTo(@NonNull Object document) {
+        for (Command command : commands) {
+            document = command.apply(document);
+        }
+        return document;
+    }
+
+    /**
+     * Add a journal command to this journal
+     *
+     * @param command The journal command
+     */
+    public void add(@NonNull Command command) {
+        this.commands.add(command);
+    }
+
+    /**
+     * Clear the journal, removing all journal entries.
+     */
+    public void clear() {
+        this.commands.clear();
+    }
+
+    /**
+     * Serialize the journal to a stream.
+     *
+     * @param out The stream to serialize to.
+     * @throws IOException if an IO exception occurs.
+     */
+    public void serialize(@NonNull OutputStream out) throws IOException {
+        for (Command command : commands) {
+            command.serialize(out);
+            out.write(0);
+            out.flush();
+        }
+    }
+
+    /**
+     * Deserialize from a byte array.
+     *
+     * @param bytes The bytes to deserialize from
+     * @return The journal
+     */
+    public static @NonNull Journal deserialize(@NonNull byte[] bytes, @NonNull Logger logger) {
+        Journal journal = new Journal();
+        int currentStart = 0;
+        for (int i = 0; i < bytes.length; i++) {
+            if (bytes[i] == 0) {
+                try {
+                    byte[] document = Arrays.copyOfRange(bytes, currentStart, i);
+                    journal.add(Command.deserialize(document));
+                    currentStart = i + 1;
+                } catch (Exception exception) {
+                    logger.e(
+                            "Could not deserialize journal entry" + Arrays.toString(bytes),
+                            exception);
+                }
+            }
+        }
+        return journal;
+    }
+
+    /**
+     * Represents a single journal command, consisting of a document path,
+     * and a value to be applied at that path level.
+     */
+    public static class Command {
+        private static final DslJson<Object> dslJson = new DslJson<>();
+
+        @NonNull
+        public final DocumentPath path;
+        @NonNull
+        public final String pathAsString;
```

Comment:
nit: this might be more useful to access the path as a string via some method on `DocumentPath` instead

Reply:
Its only purpose is for serializing the journal entry. Since DocumentPath doesn't get serialized or care about the serialized path form, I put it here.